Dynamodb cdk (#4)
Eikix/articles: blog-posts/dynamodb-cdk/dynamodb-cdk.md

---
published: false
title: 'AWS CDK & Dynamodb: this one configuration line that's costing you hundreds of dollars! 🤑 🆘'
cover_image: 'https://files.slack.com/files-pri/T7RNRLR3P-F03S45QR7PH/image.png'
description: 'AWS CDK sets your AWS Dynamodb's billing mode to "Provisioned" by default! You're potentially wasting hundreds of dollars per month because of this one line of configuration'
tags: serverless, awscdk, aws, dynamodb
series:
canonical_url:
---

> Hey there, serverless aficionadas and aficionados 😶‍🌫️! Have you ever made an oopsie 🙊 that cost your team hundreds of dollars? More than that? No way. Thousands!? First, I'm really sorry. Second, welcome to the club.

Reviewer: Ça passera peut-être mieux

```suggestion
Hey there, serverless aficionadas and aficionados 😶‍🌫️! Have you ever made an unfortunate mistake 🙊 that cost your team hundreds of dollars? More than that? No way. Thousands!? First, I'm really sorry. Second, welcome to the club.
```

## TL;DR

> When using the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) to provision a DynamoDB table, CDK sets by default the billing mode as "Provisioned" instead of pay-per-request. Therefore, Dynamodb essentially ceases to be truly serverless - remember scale-to-zero?

Reviewer: Tu peux rajouter le lien exact vers la doc où la valeur par défaut est définie

Reviewer: Pour appuyer un peu plus le propos

```suggestion
When using the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) to provision a DynamoDB table, CDK sets by default the billing mode as "Provisioned" instead of pay-per-request. Therefore, **Dynamodb essentially ceases to be truly serverless** - remember scale-to-zero?
```

> Quick fix: when instanciating your AWS Dynamodb table, make sure to explicitly set the billing mode to "pay-per-request". If you're playing around with a toy stack, you can also set the removal policy to "destroy". This means that if you remove your stack, the table will be deleted as a result (probably a bad idea to use this setting in production).

Reviewer: ```suggestion
Quick fix: when instanciating your AWS Dynamodb table, make sure to explicitly set the billing mode to "pay-per-request". If you're playing around with a toy or development stack, you can also set the removal policy to "destroy". This means that if you remove your stack, the table will be deleted as a result (probably a bad idea to use this setting in production).
```

```ts
// code/dynamodb-cdk.ts

import { App, RemovalPolicy, Stack } from '@aws-cdk/core';
import { AttributeType, Table, BillingMode } from '@aws-cdk/aws-dynamodb';

const app = new App();
const stack = new Stack(app);

const table = new Table(stack, 'MyTable', {
  partitionKey: { name: 'PK', type: AttributeType.STRING },
  sortKey: { name: 'SK', type: AttributeType.STRING },
  billingMode: BillingMode.PAY_PER_REQUEST,
  removalPolicy: RemovalPolicy.DESTROY,
});
```

### Getting acquainted with serverless as a software engineer

> One of my first missions as a rookie dev was to help design a "serverless 101 course". We aimed to help the future rookies in my team feel familiar with cloud native concepts quickly. We built a concise tutorial that goes through the basics of serverless: lambdas, API gateway, iam roles and dynamodb tables - with AWS as our cloud provider. I was pretty happy about the result! New comers could finally play in a sandbox environment with complex - and scary at first - concepts such as functions-as-a-service and cloud native databases!

Reviewer: Le terme rookie est peut-être trop lié à Theodo, `new dev` ou équivalent pourrait être mieux

```suggestion
One of my first missions as a new dev was to help design a "serverless 101 course". We aimed to help the future rookies in my team feel familiar with cloud native concepts quickly. We built a concise tutorial that goes through the basics of serverless: lambdas, API gateway, iam roles and dynamodb tables - with AWS as our cloud provider. I was pretty happy about the result! New comers could finally play in a sandbox environment with complex - and scary at first - concepts such as functions-as-a-service and cloud native databases!
```
